### ISSUE TITLE
Hot reloading of multi-module mitmproxy scripts

### ISSUE THREAD

**Issue by dkasak** [mitmproxy/mitmproxy#4253]
#### Problem Description

I have an addon consisting of multiple modules in a package. One of the modules (`main.py`) serves as an entrypoint for mitmdump which gets loaded as a script.

I would like mitmproxy to reload all of the package's modules, including the mitmproxy script defined in `main.py`, when any of them gets written to (or their mtime changes). Is this possible in current mitmproxy? I've seen some hints that it might be, but the documentation doesn't seem to mention it. When it does and when I see mentions of discussions on reloading, it is usually talking about old code which doesn't exist anymore.

Will I be able to achieve this, and if yes, how?

#### Steps to reproduce the behavior:
1. Create mitmproxy script `main.py`, which depends on `bar/baz.py` and `quux/quuz.py` which are located in the same directory.
2. Load main.py as a script in mitmproxy.
3. Do some changes to baz.py or quuz.py
4. Observe that main.py *did not* get reloaded, whereas I would like it to be.

#### System Information
```
❯ mitmproxy --version

Mitmproxy: 5.2
Python:    3.8.6
OpenSSL:   OpenSSL 1.1.1h  22 Sep 2020
Platform:  Linux-5.4.72-1-lts-x86_64-with-glibc2.2.5
```

**Comment by dkasak** [comment]
A related problem which further aggravates this is that dependency modules will not get reloaded even if you touch `main.py`. So if I make a change in `foo.py`, I have no way of having that change reflected in a running mitmdump session. Instead, I have to restart mitmdump.

Essentially, this means that all addon code *has* to live in a single module if you want hot reloading.

**Comment by mhils** [comment]
We currently only have support for hot reloading the main file (and I think we never supported more than that). Improving this probably involves 1) figuring out which modules are imported by an addon, 2) tracking changes for all of them, and 3) hot-reloading when necessary. I'd be happy to merge a PR that tackles this (preferably without adding too much complexity), but it's probably not a trivial undertaking. Do we have any idea how others (e.g. Flask) are doing this?

**Comment by dkasak** [comment]
> We currently only have support for hot reloading the main file (and I think we never supported more than that).

I guess I must have confused something. I think I got this idea from https://github.com/mitmproxy/mitmproxy/issues/1828 where there's some discussion about reloading if any *.py file in the directory changes.

I don't know how others do this, but I might investigate when I get some time. Regarding your point 1), I'd already be pretty happy to specify the modules manually too, if it makes things easier.

**Comment by dkasak** [comment]
So in case anyone else starts looking into this, there's a standard way to determine modules used by a script via [modulefinder](https://docs.python.org/3/library/modulefinder.html), but it currently [barfs on namespace packages](https://bugs.python.org/issue40350).

**Comment by Prinzhorn** [comment]
Maybe we could go an easier route and offer a `--scripts-watch` argument that accepts globs? Instead of trying to map out all the imports we put it in the hands of the user to specify which files to watch for changes (this is an advanced use-case anyway). This would be a catch-all solution and reload all the scripts.

```
mitmproxy --scripts main.py --scripts-watch 'deps/*.py, foo.py'
```

**Comment by dkasak** [comment]
Yes, I'd be fine with that. (Another possible route I thought of would be to not only support module scripts, but package scripts. Instead of specifying `--scripts module.py`, you specify `--scripts package` and all of the package's modules would be watched for changes. But as I said, I'm fine with `--scripts-watch`, just throwing the idea out there.)

> Edit: For now during dev you could probably do modd + touch main.py?

Yes, this could work. However, I also need to `importlib.reload` each module from `main.py` and I need to `touch main.py` twice before changes are visible, for reasons I'm still unclear on.

**Comment by mhils** [comment]
> Yes, this could work. However, I also need to importlib.reload each module from main.py and I need to touch main.py twice before changes are visible, for reasons I'm still unclear on.

In what order are you reloading and importing in your script? That may be an issue here.

**Comment by dkasak** [comment]
```python
# import lines here

# A

# some function definitions here

# B

addons = [MyAddon()]

# C
```

If we consider my `main.py` to have the structure above, I tried doing

```python
for m in list_of_modules:
    importlib.reload(m)
```

at point A, B and C, but the effect is the same in all three positions.

**Comment by mhils** [comment]
Have you tried repeating the import instructions after calling reload, or more generally, `foo = importlib.reload(foo)`?

**Comment by dkasak** [comment]
Ah. I did try `foo = importlib.reload(foo)`, but most of my imports were in the `from ... import ...` style, so this didn't work. Repeating the entire set of import statements after reloading works. Thanks!

**Comment by Ovsyanka** [comment]
I would like to share my realization of suggested approach: "watch for changes in files and touch main script file on every change". It is about python code you have to insert in your main script file.

<details>
  <summary>Details</summary>
  
First of all you have to use `importlib.reload(addon_name)` for all your modules you want to be reloaded. Like this:

```py
import importlib
...
import addons.addon_name
addons.addon_name = importlib.reload(addons.addon_name)
...
addons = [
    ...
    addons.addon_name.AddonClassName(),
]
```

The second part is to touch your main script (let it be `main.py`) on each change in source files. Here I watch for changes in every python related file in the project directory using [watchgod](https://github.com/samuelcolvin/watchgod) (note: not `watchdog`) package.

I use `pipx` to install `mitmproxy` and `watchgod` (as recommended in [this](https://docs.mitmproxy.org/stable/overview-installation/#installation-from-the-python-package-index-pypi) documentation section):

```sh
pipx install mitmproxy
pipx inject mitmproxy watchgod
```

Then add this code to the `main.py`:

```py
import asyncio

from watchgod import awatch
from watchgod import PythonWatcher

from pathlib import Path

class MyWatcher(PythonWatcher):
    """Use PythonWatcher, and additionaly ignore current file
    """

    def should_watch_file(self, entry: 'DirEntry') -> bool:
        is_current_file = f'./{__file__}' == entry.path
        return not is_current_file and super().should_watch_file(entry)

async def autoreload():
    ctx.log.debug("Autoreload task started")

    async for changes in awatch('.', watcher_cls=MyWatcher):
        try:
            ctx.log.debug(f"Changes detected: {changes}")
            Path(__file__).touch()
        except Exception as e:
            print("Exception in autoreload:", e)

# We have to delete tasks started at previous version of the script
for task in asyncio.all_tasks():
    if task.get_name() == 'autoreloadTask':
        ctx.log.debug("Canceling previous autoreload task")
        task.cancel()

asyncio.create_task(autoreload(), name="autoreloadTask")
```
  
</details>
